Saxon done ✓

TaliaMalia opened the pull request:
… files)

Commit:
Pull Request Draft Saxon scale added, tests are running (except .d.ts files)

musoke commented on babel.config.js:
Was this needed to get tests to run?  I don't think it should be necessary.

TaliaMalia replied:
Yes, I thought it would solve the error that I got, but the babel and jest config are actually unnecessary. I still don't know why I got that error :/

musoke replied:
Ok!

musoke replied:
I also can't tell why you got that error.  Maybe you forgot to `yarn install`?

TaliaMalia replied:
That's too bad, but thanks anyway! No, I don't think that was the problem.

musoke commented on jest.config.cjs:
This also seems like it should be unnecessary

TaliaMalia replied:
I deleted the babel config and changed the jest config to how it was before.

musoke commented on package.json:
I don't think any of these changes are needed

TaliaMalia replied:
I undid my changes in the package.json.

musoke commented on README.md:
```diff
@@ -13,6 +13,7 @@ Javascript utilities for working with rock climbing grades.
 - [x] French
 - [x] UIAA
 - [x] Ewbanks
+- [x] Saxon
```
Nice, good to remember this

musoke commented on src/__tests__/scales/saxon.ts:
```diff
@@ -0,0 +1,102 @@
+    })
+  })
+
+  describe('Invalid grade format', () => {
```
@TaliaMalia , please add a check that a grade such as 5a is disallowed.

TaliaMalia replied:
I added a test called "invalid character" that disallows to use grades < 7 to have letters from a to c.
While doing that I recognized that the saxonGradeRegex I wrote was incorrect and did not consider that the grades 1-6 are not allowed to have letters. I changed that now and everything works as it should.

musoke reviewed (changes requested):
@TaliaMalia, most of this looks good!  I have left some comments below.  In particular, the changes you made to package.json, yarn.lock, jest.config and babel.config.js aren't needed.  

Could you also add a test for an edge case grade like 5a?  It should be disallowed.

Commit:
Changes done:
Changed package.json ✓
Changed jest.config.cjs ✓
deleted babel.config.js ✓
Added a test called 'invalid character' to saxon.ts ✓

musoke commented on src/__tests__/scales/saxon.ts:
```diff
@@ -70,6 +70,11 @@ describe('Saxon', () => {
     })
+    test('invalid character', () => {
+      const invalidGrade = Saxon.getScore('5a')
+      expect(console.warn).toHaveBeenCalledWith('Unexpected grade format: 5a for grade scale Saxon')
```
good

musoke commented on src/scales/saxon.ts:
```diff
@@ -3,7 +3,7 @@ import routes from '../data/routes.json'
 import { GradeBandTypes, routeScoreToBand } from '../GradeBands'
 
-const saxonGradeRegex = /^(([1-9]|1[0-3])([a-c])?){1}(?:(\/)(([1-9]|1[0-3])([a-c])?))?$/i
+const saxonGradeRegex = /^((([7-9]|1[0-3])([a-c]))|([1-6]))$/i
```
nice

musoke committed:
fixup: remove changes to config

musoke committed:
fixup formatting

musoke committed:
fixup json formatting

musoke commented on src/scales/ewbank.ts:
```diff
@@ -26,17 +26,18 @@ const EwbankScale: GradeScale = {
       const validScore = Number.isInteger(score) ? score : Math.ceil(score)
-      return Math.min(Math.max(0, validScore), routes.length - 1)
+      const index = Math.min(Math.max(0, validScore), routes.length - 1)
+      return routes[index].ewbank
     }
 
     if (typeof score === 'number') {
-      return routes[validateScore(score)].ewbank
+      return validateScore(score)
     }
 
-    const low: string = routes[validateScore(score[0])].ewbank
-    const high: string = routes[validateScore(score[1])].ewbank
+    const low: string = validateScore(score[0])
+    const high: string = validateScore(score[1])
```
One last question: could you say again why this change was necessary?  The tests run for me without it.

TaliaMalia replied:
Yes of course. If I don't use type assertion in validateScore I get a type error **"Type 'string | number' is not assignable to type 'string'. Type 'number' is not assignable to type 'string'"**. I really can't explain to you why that only happens in ewbank or why that happens at all...

musoke commented on src/scales/uiaa.ts:
```diff
@@ -14,7 +14,7 @@ const UIAAScale: GradeScale = {
   name: GradeScales.UIAA,
   offset: 2000,
-  allowableConversionType: [GradeScales.YDS],
+  allowableConversionType: [GradeScales.YDS, GradeScales.SAXON, GradeScales.EWBANK, GradeScales.FRENCH],
```
Good catch, this was a bug we hadn't noticed.

musoke commented:
I left one more question for you.
I also pushed some changes to make ts-standard happy, so you'll have to pull before doing further edits (if any).

musoke commented:
I can't reproduce that error. Does it still happen if you undo the change?

On Wed, May 3, 2023, 03:22 Talia Aleyna Hortac ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/scales/ewbank.ts
> <https://github.com/OpenBeta/sandbag/pull/108#discussion_r1183322886>:
>
> > -    const validateScore = (score: number): number => {
> +    const validateScore = (score: number): string => {
>        const validScore = Number.isInteger(score) ? score : Math.ceil(score)
> -      return Math.min(Math.max(0, validScore), routes.length - 1)
> +      const index = Math.min(Math.max(0, validScore), routes.length - 1)
> +      return routes[index].ewbank
>      }
>
>      if (typeof score === 'number') {
> -      return routes[validateScore(score)].ewbank
> +      return validateScore(score)
>      }
>
> -    const low: string = routes[validateScore(score[0])].ewbank
> -    const high: string = routes[validateScore(score[1])].ewbank
> +    const low: string = validateScore(score[0])
> +    const high: string = validateScore(score[1])
>
> Yes of course. If I don't use type assertion in validateScore I get a type
> error *"Type 'string | number' is not assignable to type 'string'. Type
> 'number' is not assignable to type 'string'"*. I really can't explain to
> you why that only happens in ewbank or why that happens at all...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenBeta/sandbag/pull/108#discussion_r1183322886>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD7ET7DYI6QAN3J3OTLRRKDXEIBRZANCNFSM6AAAAAAXRWG54I>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

TaliaMalia commented:
Yes, the only way the code is working for me is when I run it like I committed it. If it does not have any consequences and the code is still working without my changes I can just undo the changes and push.

musoke commented:
Ok, it sounds like the version you wrote is the only one that works for both of us.  I just wish I knew why so we could avoid such inconsistencies in future.

musoke commented:
I'll squash and merge shortly

TaliaMalia commented:
I wish I knew too. I checked everything I can think of, but just couldn't figure it out :/

musoke commented:
Thank you for you contribution!

@all-contributors add @TaliaMalia  for code

TaliaMalia commented:
Your welcome!
Thank you, that you were so patient and communicative :)
